File: be-1-nestjs/src/auth/util.refresh-token.ts
// https://docs.aws.amazon.com/cognito/latest/developerguide/token-endpoint.html

import axios from 'axios';
import { config } from '@/config'

export interface RefreshAccessTokenResult {
  access_token: string,
  id_token: string,
  token_type: string, // Bearer Basic
  expires_in: number
}

export interface RefreshTokenProps {
  refreshToken: string
}

export const refreshAccessToken = async (refresh_token: string): Promise<RefreshAccessTokenResult> => {
  const { clientId, clientSecret, clientDomain, clientRegion } = config().cognito

  const url = `https://${clientDomain}.auth.${clientRegion}.amazoncognito.com/oauth2/token`
  const client_id = clientId
  const grant_type = 'refresh_token'
  const params = { grant_type, client_id, refresh_token }
  const Authorization = 'Basic ' + Buffer.from(clientId + ':' + clientSecret).toString('base64')

  const headers = {
    'Content-Type': 'application/x-www-form-urlencoded',
    'Authorization': Authorization
  }

  console.log('...COGNITO --> refreshAccessToken')
  const response = await axios.request<RefreshAccessTokenResult>({ url, params, headers, method: 'POST' })
  return response.data
}

File: be-2-flask/app/auth/util_refresh_token.py
from typing import TypedDict
import requests
from base64 import b64encode
from config import (
    COGNITO_CLIENT_DOMAIN,
    COGNITO_CLIENT_ID,
    COGNITO_CLIENT_REGION,
    COGNITO_CLIENT_SECRET,
)


class RefreshAccessTokenResult(TypedDict):
    access_token: str
    id_token: str
    token_type: str
    expires_in: str


def refresh_access_token(refresh_token: str) -> RefreshAccessTokenResult:
    url = f"https://{COGNITO_CLIENT_DOMAIN}.auth.{COGNITO_CLIENT_REGION}.amazoncognito.com/oauth2/token"
    grant_type = "refresh_token"
    data = {"grant_type": grant_type, "client_id": COGNITO_CLIENT_ID, "refresh_token": refresh_token}

    client_credentials = f"{COGNITO_CLIENT_ID}:{COGNITO_CLIENT_SECRET}"
    authorization = "Basic " + b64encode(client_credentials.encode()).decode("utf-8")

    headers = {"Content-Type": "application/x-www-form-urlencoded", "Authorization": authorization}

    print("...COGNITO --> refreshAccessToken")
    response = requests.post(url, data=data, headers=headers)
    response_data: RefreshAccessTokenResult = response.json()
    return response_data
